sw/device/lib/dif/dif_gpio.c

```diff
@@ -0,0 +1,276 @@
+ * the lower half of the actual register.
+ * @param upper_offset Offset of the masked access register that corresponds to
+ * the upper half of the actual register.
+ * @param mask Mask that identifies the bits to write to.
+ * @param val Value to write.
+ */
+static void gpio_masked_write(const dif_gpio_t *gpio, uint32_t lower_offset,
+                              uint32_t upper_offset, uint32_t mask,
+                              uint32_t val) {
+  const uint32_t mask_lower_half = mask & 0x0000FFFFU;
+  const uint32_t mask_upper_half = mask & 0xFFFF0000U;
+  if (mask_lower_half != 0) {
+    mmio_region_write32(gpio->base_addr, lower_offset,
+                        (mask_lower_half << 16) | (val & 0x0000FFFFU));
+  }
+  if (mask_upper_half != 0) {
+    mmio_region_write32(gpio->base_addr, upper_offset,
+                        mask_upper_half | ((val & 0xFFFF0000U) >> 16));
+  }
+}
+
+/**
+ * Perform a masked write to a single bit of a GPIO register.
+ *
+ * GPIO device provides masked (thread-safe) writes to some of its registers.
+ * This allows software to modify half of the bits at a time without requiring
+ * a read-modify-write. This function is guaranteed to perform only one write
+ * since it never needs to access both halves of a register.
+ *
+ * See also |gpio_masked_write()|.
+ *
+ * @param gpio GPIO instance.
+ * @param lower_offset Offset of the masked access register that corresponds to
+ * the lower half of the actual register.
+ * @param upper_offset Offset of the masked access register that corresponds to
+ * the upper half of the actual register.
+ * @param index Zero-based index of the bit to write to.
+ * @param val Value to write.
+ */
+static void gpio_masked_bit_write(const dif_gpio_t *gpio, uint32_t lower_offset,
+                                  uint32_t upper_offset, uint32_t index,
+                                  bool val) {
+  // Write to lower_offset if the bit is in the lower half, write to
+  // upper_offset otherwise.
+  const ptrdiff_t offset = (index < 16) ? lower_offset : upper_offset;
+  // Since masked access writes to half of a register, index mod 16 gives the
+  // index of the bit in the half-word mask.
+  const uint32_t mask = index_to_mask(index % 16);
+  mmio_region_write32(gpio->base_addr, offset,
+                      (mask << 16) | (val ? mask : 0U));
+}
+
+static uint32_t interrupt_rising_edge_enabled(const dif_gpio_t *gpio,
+                                              uint32_t mask) {
+  return mmio_region_read_mask32(gpio->base_addr,
+                                 GPIO_INTR_CTRL_EN_RISING_REG_OFFSET, mask, 0);
+}
+
+static uint32_t interrupt_falling_edge_enabled(const dif_gpio_t *gpio,
+                                               uint32_t mask) {
+  return mmio_region_read_mask32(gpio->base_addr,
+                                 GPIO_INTR_CTRL_EN_FALLING_REG_OFFSET, mask, 0);
+}
+
+static uint32_t interrupt_level_high_enabled(const dif_gpio_t *gpio,
```

Comment:
I agree, thanks for pointing this out. There used to be a single interrupt related function, but as I implemented the remaining interrupt types, things got complicated. I'm not sure how this helps with avoiding _incompatible setup_ though. Can you please elaborate?

Reply:
OK, just saw what you mean:`kDifGpioTriggerTypeRisingFalling`.

Reply:
I think, for this to work, we need to leave out the `enable` parameter. Otherwise, this function can be called consecutively to reach a state that we don't want.

Reply:
Updated this PR with a slightly modified version of this idea (without `enable` as I mentioned above). Please let me know if you have any comments. Thanks!